CRM/ManageLetterheads/Form/LetterheadForm.php

```diff
@@ -0,0 +1,211 @@
+<?php
+
+use CRM_ManageLetterheads_BAO_LetterheadAvailability as LetterheadAvailability;
+
+/**
+ * Create Letterhead Form class.
+ */
+class CRM_ManageLetterheads_Form_LetterheadForm extends CRM_Core_Form {
+
+  /**
+   * @var string
+   *   The ID for the letterhead. Used when updating.
+   */
+  private $_letterheadId;
+
+  /**
+   * Adds the form fields and buttons.
+   */
+  public function buildQuickForm() {
+    $this->addFormElements();
+    $this->setDefaultWeight();
+    $this->assign('elementNames', $this->getRenderableElementNames());
+    $this->preventAjaxSubmit();
+    parent::buildQuickForm();
+  }
+
+  /**
+   * Stores the letterhead ID internally, if provided.
+   */
+  public function preProcess() {
+    $this->_letterheadId = CRM_Utils_Request::retrieve('id', 'Positive', $this);
+  }
+
+  /**
+   * Populates the letterhead fields when using the update form.
+   *
+   * {@inheritDoc}
+   */
+  public function setDefaultValues() {
+    $hasDefaultValues = !empty($this->defaultValues);
+    $isUpdateAction = $this->_action & CRM_Core_Action::UPDATE;
+
+    if ($hasDefaultValues || !$isUpdateAction) {
+      return;
+    }
+
+    $letterhead = civicrm_api3('Letterhead', 'getsingle', [
+      'id' => $this->_letterheadId,
+    ]);
+
+    $letterhead['available_for'] = $this->getAvailableForValuesAsKeys(
+      $letterhead['available_for']
+    );
+
+    $this->defaultValues = $letterhead;
+
+    return $this->defaultValues;
+  }
+
+  /**
+   * Handles the letterhead creation and update process.
+   */
+  public function postProcess() {
+    $formValues = $this->controller->exportValues($this->_name);
+    $isUpdateAction = $this->_action & CRM_Core_Action::UPDATE;
+
+    if ($isUpdateAction) {
+      $formValues['id'] = $this->_letterheadId;
+    }
+
+    $this->createLetterhead($formValues);
+    $this->addSuccessMessage();
+    $this->redirectToLetterheadsListPage();
+  }
+
+  /**
+   * Flips an array of Available For values as keys of the array.
+   *
+   * This is done because they are stored as [a, b, c], where a, b, and are
+   * the Available For option values, but the form understands it as
+   * [a => 1, b => 0, c => 1] where 1 or 0 means checked/uncheked.
+   *
+   * @param array $availableFor
+   *   An array Available For option values.
+   * @return array
+   */
+  private function getAvailableForValuesAsKeys(array $availableFor) {
```

Comment:
Is this function doing what we want it to do?
First of all do we need to flip the `$availableFor` seeing that it contains an array of available for option values?

Also the array_map returns `1` for all values.

Reply:
@tunbola this is why I created a function, to try to explain this well. Might have to explain it better. The thing is that the checkboxes work like this:

```php
['value' => 'checked']
```

Where `value` is the "Available For" option value, and `checked` is either `1` or `0`. `1` means the particular value is checked. Since all the given options must be checked we always use `1`.

Reply:
@reneolivo I see. 

In that case, we can try this then

`array_fill_keys($availableFor, 1);`
We might not even need this function and we can just use this directly in `setDefaultValues`

Reply:
Thanks a lot @tunbola . Changed the implementation. Should I remove the function altogether? I was thinking of leaving it to have an explanation about the checkbox behaviour.

Let me know if you want me to remove it.

Reply:
It's fine @reneolivo . We can leave the function.